### PR TITLE
Tutorial on KerasModel and TorchModel

### PR DESCRIPTION
**Pull request by peastman**
I didn't give the tutorial a number, since we're going to reorder all of them anyway.  The change in `base_classes.py` was needed to make it work with PyTorch.

### CODE REVIEW

**Comment by nd-02110114** [comment]
This tutorial is really useful for many users and the contents are good for me!

But, I just have one question.

I seem the roc_auc_score (both train and test) about `ClassificationModel` is too worse. 
```
training set score: {'roc_auc_score': 0.4268401201369002}
test set score: {'roc_auc_score': 0.22269021739130435}
```

Is this an expected situation? I tried to use RandomForest, but I could get more suitable score.

```python
import sklearn

sklearn_model = sklearn.ensemble.RandomForestRegressor(n_estimators=300)
model = dc.models.SklearnModel(sklearn_model)
tasks, datasets, transformers = dc.molnet.load_bace_classification(feturizer='ECFP', split='scaffold')
train_dataset, valid_dataset, test_dataset = datasets
model.fit(train_dataset)
metric = dc.metrics.Metric(dc.metrics.roc_auc_score)
print('train set score:', model.evaluate(train_dataset, [metric]))
print('test set score:', model.evaluate(test_dataset, [metric]))
```

The result

```
training set score: {'roc_auc_score': 0.9994859258224488}
test set score: {'roc_auc_score': 0.8529891304347825}
```

And, then I checked the loss values of `ClassificationModel`, but I seem these are strange... these are increasing...

```python
import deepchem as dc
import tensorflow as tf

class ClassificationModel(tf.keras.Model):
    
    def __init__(self, n_classes):
        super(ClassificationModel, self).__init__()
        self.dense1 = tf.keras.layers.Dense(1000, activation='relu')
        self.dense2 = tf.keras.layers.Dense(n_classes)

    def call(self, inputs, training=False):
        y = self.dense1(inputs)
        if training:
            y = tf.nn.dropout(y, 0.5)
        logits = self.dense2(y)
        output = tf.nn.softmax(logits)
        return output, logits

keras_model = ClassificationModel(2)
output_types = ['prediction', 'loss']
model = dc.models.KerasModel(keras_model, dc.models.losses.SoftmaxCrossEntropy(), output_types=output_types)
tasks, datasets, transformers = dc.molnet.load_bace_classification(feturizer='ECFP', split='scaffold')
train_dataset, valid_dataset, test_dataset = datasets
for _ in range(10):
    loss = model.fit(train_dataset, nb_epoch=1)
    print(loss)
```

The result (Increasing....)

```
1.1389728692861705
1.1541260939378004
1.205811353830191
1.3021155137282152
1.4290218353271484
1.6480687948373647
1.7665462493896484
1.6819359064102173
2.333866706261268
2.5435421283428488
```

However, I seem regression model achieve good convergence.
```python
tasks, datasets, transformers = dc.molnet.load_delaney(featurizer='ECFP', splitter='random')
train_dataset, valid_dataset, test_dataset = datasets
keras_model = tf.keras.Sequential([
    tf.keras.layers.Dense(1000, activation='relu'),
    tf.keras.layers.Dropout(rate=0.5),
    tf.keras.layers.Dense(1)
])
model = dc.models.KerasModel(keras_model, dc.models.losses.L2Loss())
for _ in range(10):
    loss = model.fit(train_dataset, nb_epoch=1)
    print(loss)
``` 

The result (decereasing, good convergence)

```
0.6786340713500977
0.34340848922729494
0.21966273784637452
0.16352051496505737
0.1269517183303833
0.10376811027526855
0.09067392945289612
0.08337926268577575
0.07561246752738952
0.06793087124824523
```

I faced similar situation for GATModel and CGCNNModel, so I want to confirm whether this behavior of classification model is proper or not.

**Comment by peastman** [comment]
I ran into the same situation while investigating tutorial 12 (https://github.com/deepchem/deepchem/issues/2112#issuecomment-684022015).  Using BACE with scaffold splitting gives really bad results.  It's possible we could get better results by optimizing the hyperparameters.

**Comment by peastman** [comment]
I don't really understand what's going on.  As training progresses, the loss goes up and down mostly at random.  The ROC AUC also goes up and down.  I train a bit and the score is terrible.  I train a bit more and it gets much better (on both training and test sets).  I train some more and it's terrible again.

Normally I would think this just meant the learning rate was too high, but in fact decreasing the learning rate doesn't help.  The behavior remains the same.  Furthermore, there seems to be no correlation between the loss and the ROC AUC!  Sometimes the loss is very high but the test score is excellent.  Other times the loss is low but the AUC is terrible.

**Comment by rbharath** [comment]
So taking a look, I'm able to get the following snippet with `dc.models.MultitaskClassifier` on BACE with scaffold split to work:

```
import sklearn
import sklearn.ensemble
import deepchem as dc

tasks, datasets, transformers = dc.molnet.load_bace_classification(feturizer='ECFP', split='scaffold')
model = dc.models.MultitaskClassifier(len(tasks), 1024)
train_dataset, valid_dataset, test_dataset = datasets
model.fit(train_dataset, nb_epoch=5)
metric = dc.metrics.Metric(dc.metrics.roc_auc_score)
print('train set score:', model.evaluate(train_dataset, [metric]))
print('test set score:', model.evaluate(test_dataset, [metric]))
```
This prints
```
train set score: {'roc_auc_score': 0.9665293008311797}
test set score: {'roc_auc_score': 0.8201086956521739}
```

On the other hand, using `ClassificationModel` defined as above gives me the oscillating training. Here's a printout of losses and train/test scores for 10 epochs:

```
Starting epoch 0
1.1495058353130634
train set score: {'roc_auc_score': 0.49928616330236775}
test set score: {'roc_auc_score': 0.5902173913043478}
Starting epoch 1
1.1507729016817534
train set score: {'roc_auc_score': 0.4008968359293148}
test set score: {'roc_auc_score': 0.49076086956521736}
Starting epoch 2
1.1866917243370643
train set score: {'roc_auc_score': 0.30412795976810786}
test set score: {'roc_auc_score': 0.44003623188405794}
Starting epoch 3
1.2968499110295222
train set score: {'roc_auc_score': 0.37106097646154923}
test set score: {'roc_auc_score': 0.34130434782608693}
Starting epoch 4
1.4362192887526293
train set score: {'roc_auc_score': 0.32344625270657257}
test set score: {'roc_auc_score': 0.40108695652173915}
Starting epoch 5
1.5846560551570013
train set score: {'roc_auc_score': 0.6863728434727946}
test set score: {'roc_auc_score': 0.6070652173913044}
Starting epoch 6
1.8062804295466497
train set score: {'roc_auc_score': 0.5826709506181462}
test set score: {'roc_auc_score': 0.5105072463768117}
Starting epoch 7
1.8799299001693726
train set score: {'roc_auc_score': 0.2573653698400503}
test set score: {'roc_auc_score': 0.40398550724637683}
Starting epoch 8
2.1171229435847354
train set score: {'roc_auc_score': 0.3809303625061117}
test set score: {'roc_auc_score': 0.3606884057971015}
Starting epoch 9
2.477137932410607
train set score: {'roc_auc_score': 0.4777271774813159}
test set score: {'roc_auc_score': 0.37282608695652175}
```

I'm not sure what the difference is. BACE with scaffold, the dataset itself, looks to be OK, but perhaps it's something `ClassificationModel` is missing?

**Comment by nd-02110114** [comment]
> I'm not sure what the difference is. BACE with scaffold, the dataset itself, looks to be OK, but perhaps it's something ClassificationModel is missing?

I agree. I think this is not related to dataset itself and hyperparameters.

**Comment by peastman** [comment]
I think it was a problem in the shape of outputs.  Since it only has two classes I switched it to outputting sigmoid rather than softmax and now it works well.

I also removed an inappropriate warning message.  Every time I imported deepchem it displayed the message, "HuggingFace transformers is not available. Please install using 'pip install transformers' to use the SmilesTokenizer."  Warnings about optional dependencies shouldn't be displayed if you don't use any feature that requires them.  @seyonechithrananda you might want to add a warning message when the user tries to use SmilesTokenizer.

File: deepchem/feat/base_classes.py [hunk 1]
```diff
@@ -262,7 +262,7 @@ def featurize(self, molecules, log_every_n=1000):
             "Failed to featurize datapoint %d. Appending empty array")
         features.append(np.array([]))
 
-    features = np.asarray(features)
```
Comment (rbharath) [inline deepchem/feat/base_classes.py:265]: I think this line might be causing a lot of the Travis failures we see since for graph conv featurizations, we pass around `np.ndarray` objects with `dtype=object` (for `ConvMol`, `GraphData` etc).

**Review by rbharath — COMMENTED** [review]
I think the change to `base_classes.py` is causing a lot of the travis failures we're seeing. Once we figure those out, this should be good to merge!

File: deepchem/feat/base_classes.py (hunk 1 above)
Comment (peastman) [reply deepchem/feat/base_classes.py]: I reverted the change.  Let's see if I can find a more focused way of fixing the PyTorch error.

**Comment by peastman** [comment]
All tests are passing.

**Comment by seyonechithrananda** [comment]
> I think it was a problem in the shape of outputs. Since it only has two classes I switched it to outputting sigmoid rather than softmax and now it works well.
> 
> I also removed an inappropriate warning message. Every time I imported deepchem it displayed the message, "HuggingFace transformers is not available. Please install using 'pip install transformers' to use the SmilesTokenizer." Warnings about optional dependencies shouldn't be displayed if you don't use any feature that requires them. @seyonechithrananda you might want to add a warning message when the user tries to use SmilesTokenizer.

Thanks for noting this! Will remove the logger command and raise a PR
